hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -891,50 +902,123 @@ private boolean innerRename(Path source, Path dest)
       }
 
       List<DeleteObjectsRequest.KeyVersion> keysToDelete = new ArrayList<>();
+      List<DeleteObjectsRequest.KeyVersion> dirKeysToDelete = new ArrayList<>();
       if (dstStatus != null && dstStatus.isEmptyDirectory() == Tristate.TRUE) {
         // delete unnecessary fake directory.
         keysToDelete.add(new DeleteObjectsRequest.KeyVersion(dstKey));
       }
 
-      Path parentPath = keyToPath(srcKey);
-      RemoteIterator<LocatedFileStatus> iterator = listFilesAndEmptyDirectories(
-          parentPath, true);
-      while (iterator.hasNext()) {
-        LocatedFileStatus status = iterator.next();
-        long length = status.getLen();
-        String key = pathToKey(status.getPath());
-        if (status.isDirectory() && !key.endsWith("/")) {
-          key += "/";
-        }
-        keysToDelete
-            .add(new DeleteObjectsRequest.KeyVersion(key));
-        String newDstKey =
-            dstKey + key.substring(srcKey.length());
-        copyFile(key, newDstKey, length);
-
-        if (hasMetadataStore()) {
-          // with a metadata store, the object entries need to be updated,
-          // including, potentially, the ancestors
-          Path childSrc = keyToQualifiedPath(key);
-          Path childDst = keyToQualifiedPath(newDstKey);
-          if (objectRepresentsDirectory(key, length)) {
-            S3Guard.addMoveDir(metadataStore, srcPaths, dstMetas, childSrc,
-                childDst, username);
+      // A blocking queue that tracks all objects that need to be deleted
+      BlockingQueue<Optional<DeleteObjectsRequest.KeyVersion>> deleteQueue = new ArrayBlockingQueue<>(
+              (int) Math.round(MAX_ENTRIES_TO_DELETE * 1.5));
+
+      // Used to track if the delete thread was gracefully shutdown
+      boolean deleteFutureComplete = false;
+      FutureTask<Void> deleteFuture = null;
+
+      try {
+        // Launch a thread that will read from the deleteQueue and batch delete any files that have already been copied
+        deleteFuture = new FutureTask<>(() -> {
+          while (true) {
+            while (keysToDelete.size() < MAX_ENTRIES_TO_DELETE) {
+              Optional<DeleteObjectsRequest.KeyVersion> key = deleteQueue.take();
+
+              // The thread runs until is is given an EOF message (an Optional#empty())
+              if (key.isPresent()) {
```

Comment:
Use of Optional will make this harder to backport to Java 7 branches, no?   Looks like you are using it for an EOF marker only? You could also create a static final KeyVersion instance and use a reference equality comparison to detect that it is the special EOF version?

Reply:
I'm making the leap to Java 8 with the committer and retry logic. Doesn't mean we should jump to using Optional just because we can. Put differently: if you are going to just use it as a fancy null, it's overkill. Key benefit is that you and use map & foreach, but there we are crippled by java's checked exceptions stopping us throwing IOEs

Reply:
I removed the usage of `Optional`. Using a `private static final DeleteObjectsRequest.KeyVersion END_OF_KEYS_TO_DELETE = new DeleteObjectsRequest.KeyVersion(null, null);` as the EOF instead.